test/common/config/registry_test.cc

```diff
@@ -182,6 +193,50 @@ TEST(RegistryTest, TestDoubleRegistrationByName) {
                             "Double registration for name: 'testing.published.test'");
 }
 
+class TestVersionedWithDeprecatedNamesFactoryAndAdditionalCategory : public PublishedFactory {
+public:
+  std::string category() const override { return "testing.published.additional.category"; }
+  std::string name() const override {
+    return "testing.published.versioned.instead_name_and_category";
+  }
+};
+
+TEST(RegistryTest, DEPRECATED_FEATURE_TEST(VersionedWithDeprecatedNamesFactoryAndNewCategory)) {
+  TestVersionedWithDeprecatedNamesFactoryAndAdditionalCategory test;
+
+  // Check the category is not registered
+  EXPECT_FALSE(Registry::FactoryCategoryRegistry::isRegistered(test.category()));
+
+  auto factory =
+      Registry::RegisterFactory<TestVersionedWithDeprecatedNamesFactoryAndAdditionalCategory,
+                                PublishedFactory>(
+          FACTORY_VERSION(0, 0, 1, {{"build.kind", "private"}}),
+          {"testing.published.versioned.deprecated_name_and_category"});
+
+  // Check the category now registered
+  EXPECT_TRUE(Registry::FactoryCategoryRegistry::isRegistered(test.category()));
+
+  const auto& factories = Envoy::Registry::FactoryCategoryRegistry::registeredFactories();
+
+  auto version =
+      factories.find("testing.published.additional.category")
+          ->second->getFactoryVersion("testing.published.versioned.instead_name_and_category");
+  EXPECT_TRUE(version.has_value());
```

Comment:
This should be an ASSERT since if this fails, the `.value()` calls below don't make any sense. Alternatively, you can use the `testing::Optional` matcher with `testing::AllOf` and `testing::Property` to combine all the EXPECT_EQ into a single EXPECT, but that's probably less readable than what you have here.